### DISCUSSION TITLE
Necessity for CeedElemTopology?

### DISCUSSION THREAD

**Discussion by NamjaeChoi** [CEED/libCEED#1679] [Q&A] [answered]
Hello,

I'm new to libCEED and studying some APIs and examples. It looks like libCEED is agnostic to element types, as users can explicitly provide the shape functions and quadrature sets through `CeedBasisCreateH1` with arbitrary ordering and number of nodes. Element restriction also doesn't seem to require this information as the offsets to the L-vector can be set by users. What's the reason for `CeedBasisCreateH1` to receive `CeedElemTopology` as argument?

Plus, is it users' responsibility to implement subdomain restriction, i.e. the P operator?

Thank you.

**Comment by jeremylt** [comment, marked as answer]
The CeedElemRestriction $\mathscr{E}$ is purely local to the MPI rank. All global MPI communication such as parallel gather scatter operations P are the responsibility of user code. We have examples of using PETSc, MFEM, deal.II, or Nek5000 managing that portion of the work.

For the topology, internally we're currently only using the topology to tell us the dimension. Eventually it would be nice to more efficiently support topologies that have a mixed tensor product decomposition.

**Reply by NamjaeChoi** [reply]
Thank you for the clarification.
